src/core/meshcop/border_agent.hpp

```diff
@@ -387,13 +410,116 @@ class BorderAgent : public InstanceLocator, private NonCopyable
 
     static Coap::Message::Code CoapCodeFromError(Error aError);
 
+    class MeshCopTxtEncoder : public InstanceLocator
+    {
+    public:
+        explicit MeshCopTxtEncoder(Instance &aInstance)
+            : InstanceLocator(aInstance)
+        {
+        }
+
+        enum : uint8_t
+        {
+            kConnectionModeDisabled = 0,
+            kConnectionModePskc     = 1,
+            kConnectionModePskd     = 2,
+            kConnectionModeVendor   = 3,
+            kConnectionModeX509     = 4,
+        };
+
+        enum : uint8_t
+        {
+            kThreadIfStatusNotInitialized = 0,
+            kThreadIfStatusInitialized    = 1,
+            kThreadIfStatusActive         = 2,
+        };
+
+        enum : uint8_t
+        {
+            kThreadRoleDisabledOrDetached = 0,
+            kThreadRoleChild              = 1,
+            kThreadRoleRouter             = 2,
+            kThreadRoleLeader             = 3,
+        };
+
+        enum : uint8_t
+        {
+            kAvailabilityInfrequent = 0,
+            kAvailabilityHigh       = 1,
+        };
+
+        struct StateBitmap
+        {
+            uint32_t mConnectionMode : 3;
+            uint32_t mThreadIfStatus : 2;
+            uint32_t mAvailability : 2;
+            uint32_t mBbrIsActive : 1;
+            uint32_t mBbrIsPrimary : 1;
+            uint32_t mThreadRole : 2;
+            uint32_t mEpskcSupported : 1;
+
+            StateBitmap(void)
+                : mConnectionMode(0)
+                , mThreadIfStatus(0)
+                , mAvailability(0)
+                , mBbrIsActive(0)
+                , mBbrIsPrimary(0)
+                , mThreadRole(kThreadRoleDisabledOrDetached)
+                , mEpskcSupported(0)
+            {
+            }
+
+            uint32_t ToUint32(void) const
+            {
+                uint32_t bitmap = 0;
+
+                bitmap |= mConnectionMode << 0;
+                bitmap |= mThreadIfStatus << 3;
+                bitmap |= mAvailability << 5;
+                bitmap |= mBbrIsActive << 7;
+                bitmap |= mBbrIsPrimary << 8;
+                bitmap |= mThreadRole << 9;
+                bitmap |= mEpskcSupported << 11;
+                return bitmap;
+            }
+        };
+        Error EncodeTxtData(uint8_t *aOutput, uint16_t &aLength);
+
+    private:
+        Error AppendTxtEntryTo(Appender &aAppender, const char *aKey, const void *aValue, uint16_t aValueLength);
+
+        template <typename ObjectType>
+        Error AppendTxtEntryTo(Appender &aAppender, const char *aKey, const ObjectType &aObject)
+        {
+            static_assert(!TypeTraits::IsPointer<ObjectType>::kValue, "ObjectType must not be a pointer");
+
```

Comment:
It is nice adding a template specialization for `NameData` :)

To be super-safe, how about adding the following `static_assert()` here (I am always concerned of older toolchains misbehave with template specialization, this way we can detect it):

```c++
     static_assert(!TypeTraits::IsSame<ObjectType, NameData>::kValue, "ObjectType must not be `NameData`");
```

Reply:
Good point! Added the check.